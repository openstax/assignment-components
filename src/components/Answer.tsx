import cn from 'classnames';
import { ReactNode } from 'react';
import { ALPHABET, isAnswerChecked, isAnswerCorrect, isAnswerIncorrect } from '../utils';
import { Answer as AnswerType, ChosenAnswer, ID } from '../types';
import { Content } from './Content';
import { SimpleFeedback } from './Feedback';

export interface AnswerProps {
  answer: AnswerType;
  iter: number;
  qid: ID;
  type: 'teacher-review' | 'teacher-preview' | 'student' | 'student-mpp';
  hasCorrectAnswer?: boolean;
  onChangeAnswer: () => void;
  disabled: boolean;
  chosenAnswer: ChosenAnswer;
  correctAnswerId?: ID | null;
  incorrectAnswerId?: ID;
  onKeyPress?: () => void;
  answered_count?: number;
  correctIncorrectIcon?: ReactNode,
  radioBox?: ReactNode;
  contentRenderer?: JSX.Element;
  show_all_feedback: boolean;
}

export const Answer = (props: AnswerProps) => {
  const {
    type,
    iter,
    answer,
    disabled,
    onKeyPress,
    qid,
    chosenAnswer,
    correctAnswerId,
    incorrectAnswerId,
    hasCorrectAnswer,
    answered_count,
    contentRenderer,
    show_all_feedback
  } = props;

  let body, feedback, selectedCount;

  const isChecked = isAnswerChecked(answer, chosenAnswer);
  const isCorrect = isAnswerCorrect(answer, correctAnswerId);
  const isIncorrect = isAnswerIncorrect(answer, incorrectAnswerId);

  const correctIncorrectIcon = (
    <div className="correct-incorrect">
      {isCorrect && props.correctIncorrectIcon}
    </div>
  );

  const classes = cn('answers-answer', {
    'disabled': disabled,
    'answer-checked': isChecked,
    'answer-correct': isCorrect && type !== 'student-mpp',
    'answer-incorrect': incorrectAnswerId && isAnswerIncorrect(answer, incorrectAnswerId),
  });

  let ariaLabel = `${isChecked ? 'Selected ' : ''}Choice ${ALPHABET[iter]}`;
  // somewhat misleading - this means that there is a correct answer,
  // not necessarily that this answer is correct
  if (hasCorrectAnswer) {
    ariaLabel += `(${isCorrect ? 'Correct' : 'Incorrect'} Answer)`;
  }
  ariaLabel += ':';

  let onChangeAnswer, radioBox;

  if (!hasCorrectAnswer
    && (type !== 'teacher-review')
    && (type !== 'teacher-preview')
    && (type !== 'student-mpp')) {
    ({ onChangeAnswer } = props);
  }

  if (onChangeAnswer) {
    radioBox = (
      <input
        type="radio"
        className="answer-input-box"
        checked={isChecked}
        id={`${qid}-option-${iter}`}
        name={`${qid}-options`}
        onChange={onChangeAnswer}
        disabled={disabled}
      />
    );
  }

  if (show_all_feedback && answer.feedback_html) {
    feedback = (
      <SimpleFeedback key="question-mc-feedback" contentRenderer={contentRenderer}>
        {answer.feedback_html}
      </SimpleFeedback>
    );
  }

  if (type === 'teacher-review') {
    let percent = 0;
    if (answer.selected_count && answered_count) {
      percent = Math.round((answer.selected_count / answered_count) * 100);
    }
    selectedCount = (
      <span
        className="selected-count"
        data-percent={`${percent}`}
      >
        {answer.selected_count}
      </span>
    );

    body = (
      <div className="review-wrapper">
        <div className={cn('review-count', { 'green': isCorrect, 'red': !isCorrect })}>
          {selectedCount}
          <span className={cn('letter', { 'green': isCorrect, 'red': !isCorrect })}>
            {ALPHABET[iter]}
          </span>
        </div>

        <div className="answer-answer">
          <Content className="answer-content" component={contentRenderer} html={answer.content_html} />
          {feedback}
        </div>
      </div>
    );
  } else {
    body = (
      <>
        {type === 'teacher-preview' && correctIncorrectIcon}
        {selectedCount}
        {radioBox}
        <label
          onKeyPress={onKeyPress}
          htmlFor={`${qid}-option-${iter}`}
          className="answer-label">
          <span className="answer-letter-wrapper">
            <button
              onClick={onChangeAnswer}
              aria-label={ariaLabel}
              className="answer-letter"
              disabled={disabled || isIncorrect}
              data-test-id={`answer-choice-${ALPHABET[iter]}`}
            >
              {ALPHABET[iter]}
            </button>
          </span>
          <div className="answer-answer">
            <Content className="answer-content" component={contentRenderer} html={answer.content_html} />
            {feedback}
          </div>
        </label>
      </>
    );
  }

  return (
    <div className="openstax-answer">
      <section role="region" className={classes}>
        {body}
      </section>
    </div>
  );
}
Answer.displayName = 'OSAnswer';
