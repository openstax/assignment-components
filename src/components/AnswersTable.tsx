import { defaultAnswerType } from "../constants";
import { Answer as AnswerType, AnswerDisplayType, ID, Question } from "src/types";
import { Answer } from "./Answer";
import { Feedback } from "./Feedback";

export interface AnswersTableProps {
  question: Question;
  type?: AnswerDisplayType;
  answer_id?: ID;
  correct_answer_id?: ID;
  incorrectAnswerId?: ID;
  answerIdOrder?: ID[],
  feedback_html: string;
  correct_answer_feedback_html?: string;
  answered_count?: number;
  show_all_feedback?: boolean;
  onChangeAnswer: () => void;
  hideAnswers: boolean;
  hasCorrectAnswer?: boolean;
  onChangeAttempt?: () => void;
  choicesEnabled?: boolean;
  onKeyPress?: () => void;
  contentRenderer?: JSX.Element;
}

export const AnswersTable = (props: AnswersTableProps) => {
  let idCounter = 0;

  const {
    question, hideAnswers, type = defaultAnswerType, answered_count, choicesEnabled, correct_answer_id,
    incorrectAnswerId, answer_id, feedback_html, correct_answer_feedback_html,
    show_all_feedback = false, hasCorrectAnswer, onChangeAnswer, onKeyPress, answerIdOrder
  } = props;
  if (hideAnswers) { return null; }

  const { id } = question;

  const feedback: { index: number, html: string }[] = [];
  const instructions = undefined;

  const chosenAnswer = [answer_id];

  const sortedAnswersByIdOrder = (idOrder: ID[]) => {
    const { answers } = question;
    return answers.slice().sort((a, b) => idOrder.indexOf(a.id) - idOrder.indexOf(b.id));
  }

  const questionAnswerProps = {
    qid: id || `auto-${idCounter++}`,
    correctAnswerId: correct_answer_id,
    incorrectAnswerId,
    hasCorrectAnswer,
    chosenAnswer,
    onChangeAnswer: onChangeAnswer,
    type,
    answered_count,
    disabled: !choicesEnabled,
    show_all_feedback,
    onKeyPress
  };

  const answers = answerIdOrder ? sortedAnswersByIdOrder(answerIdOrder) : question.answers;

  const answersHtml = answers.map((answer, i) => {
    const additionalProps: { answer: AnswerType, iter: number, key: string }
      = { answer, iter: i, key: `${questionAnswerProps.qid}-option-${i}` };
    const answerProps = Object.assign({}, additionalProps, questionAnswerProps);

    if (answer.id === incorrectAnswerId && feedback_html) {
      feedback.push({ index: i, html: feedback_html })
    } else if (answer.id === correct_answer_id && correct_answer_feedback_html) {
      feedback.push({ index: i, html: correct_answer_feedback_html })
    }

    return (
      <Answer {...answerProps} />
    );
  });

  feedback.forEach((item, i) => {
    const spliceIndex = item.index + i + 1;
    answersHtml.splice(spliceIndex, 0, (
      <Feedback key={spliceIndex} contentRenderer={props.contentRenderer}>
        {item.html}
      </Feedback>
    ));
  });

  return (
    <div className="answers-table">
      {instructions}
      {answersHtml}
    </div>
  );
}
