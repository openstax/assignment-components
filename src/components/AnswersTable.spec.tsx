import { AnswersTable, AnswersTableProps } from './AnswersTable';
import renderer from 'react-test-renderer';
import { answerContent } from '../test/fixtures';
import { Answer } from './Answer';

describe('AnswersTable', () => {
  let props: AnswersTableProps;

  beforeEach(() => {
    props = {
      question: {
        id: '1',
        stem_html: '',
        collaborator_solutions: [],
        formats: [],
        stimulus_html: '',

        answers: [{
          id: '1',
          correctness: undefined,
          content_html: 'True'
        }, {
            id: '2',
            correctness: undefined,
            content_html: 'False'
        }],
      },
      type: 'student',
      answer_id: '',
      correct_answer_id: '',
      feedback_html: '',
      correct_answer_feedback_html: '',
      answered_count: 0,
      show_all_feedback: false,
      onChangeAnswer: jest.fn(),
      hideAnswers: false,
      hasCorrectAnswer: false,
      onChangeAttempt: jest.fn(),
      choicesEnabled: false,
    };
  });

  it('matches tutor teacher-preview snapshot', () => {
    props.question.answers.forEach((answer, i) => answer.content_html = answerContent[i]);
    const tree = renderer.create(
      <AnswersTable {...props} type="teacher-preview" onKeyPress={() => null} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correct answer feedback', () => {
    const tree = renderer.create(
      <AnswersTable {...props}
        answer_id="1"
        correct_answer_id="1"
        correct_answer_feedback_html="Feedback"
        hasCorrectAnswer={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders incorrect answer feedback', () => {
    const tree = renderer.create(
      <AnswersTable {...props}
        answer_id="1"
        correct_answer_id="2"
        incorrectAnswerId="1"
        feedback_html="Feedback"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('hides answers', () => {
    const tree = renderer.create(
      <AnswersTable {...props} hideAnswers={true} />
    ).toJSON();
    expect(tree).toBeNull();
  });

  it('generates an id if missing', () => {
    const tree = renderer.create(
      <AnswersTable {...props} question={{...props.question, id: ''}} />
    );
    expect(tree.root.findAllByProps({ qid: 'auto-0' }).length).toBe(2);
  });

  it('defaults type and show_all_feedback', () => {
    const tree = renderer.create(
      <AnswersTable {...props} type={undefined} show_all_feedback={undefined} />
    );
    const allProps = tree.root.findAllByType(Answer).map((a) => a.props);
    expect(allProps.map((props) => props['type'])).toEqual(['student', 'student']);
    expect(allProps.map((props) => props['show_all_feedback'])).toEqual([false, false]);
  });

  it('renders teacher-preview type', () => {
    const type = 'teacher-preview';
    const tree = renderer.create(
      <AnswersTable {...props} question={{...props.question, id: ''}} type={type} />
    );
    expect(tree.root.findAllByType(Answer).map((a) => a.props['type'])).toEqual([type, type]);
  });

  it('sorts by given ID order', () => {
    const tree = renderer.create(
      <AnswersTable {...props} answerIdOrder={['2', '1']} />
    );
    expect(tree.root.findAllByType(Answer).map((a) => a.props.answer.id)).toEqual(['2', '1']);
    expect(tree).toMatchSnapshot();
  });
});
