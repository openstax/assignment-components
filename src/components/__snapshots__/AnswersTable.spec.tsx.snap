// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswersTable matches tutor teacher-preview snapshot 1`] = `
<div
  className="answers-table"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
      role="region"
    >
      <div
        className="correct-incorrect"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
        onKeyPress={[Function]}
      >
        <span
          className="answer-letter-wrapper"
        >
          <button
            aria-label="Choice a:"
            className="answer-letter"
            data-test-id="answer-choice-a"
            disabled={true}
          >
            a
          </button>
        </span>
        <div
          className="answer-answer"
        >
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
      role="region"
    >
      <div
        className="correct-incorrect"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
        onKeyPress={[Function]}
      >
        <span
          className="answer-letter-wrapper"
        >
          <button
            aria-label="Choice b:"
            className="answer-letter"
            data-test-id="answer-choice-b"
            disabled={true}
          >
            b
          </button>
        </span>
        <div
          className="answer-answer"
        >
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ol>
  <li>matter is moving at speeds greater than roughly 1% the speed of light,</li>
  <li>objects are large enough to be seen with the naked eye, and</li>
  <li>there is the involvement of a strong gravitational field</li>
</ol>
",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;

exports[`AnswersTable renders correct answer feedback 1`] = `
<div
  className="answers-table"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled answer-checked answer-correct"
      role="region"
    >
      <label
        className="answer-label"
        htmlFor="1-option-0"
      >
        <span
          className="answer-letter-wrapper"
        >
          <button
            aria-label="Selected Choice a(Correct Answer):"
            className="answer-letter"
            data-test-id="answer-choice-a"
            disabled={true}
          >
            a
          </button>
        </span>
        <div
          className="answer-answer"
        >
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "True",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <aside
    className="question-feedback bottom"
  >
    <div
      aria-label="Answer Feedback"
      className="arrow"
    />
    <aside>
      <div
        className="question-feedback-content has-html"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Feedback",
          }
        }
      />
    </aside>
  </aside>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
      role="region"
    >
      <label
        className="answer-label"
        htmlFor="1-option-1"
      >
        <span
          className="answer-letter-wrapper"
        >
          <button
            aria-label="Choice b(Incorrect Answer):"
            className="answer-letter"
            data-test-id="answer-choice-b"
            disabled={true}
          >
            b
          </button>
        </span>
        <div
          className="answer-answer"
        >
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "False",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;

exports[`AnswersTable renders incorrect answer feedback 1`] = `
<div
  className="answers-table"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled answer-checked answer-incorrect"
      role="region"
    >
      <input
        checked={true}
        className="answer-input-box"
        disabled={true}
        id="1-option-0"
        name="1-options"
        onChange={[MockFunction]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
      >
        <span
          className="answer-letter-wrapper"
        >
          <button
            aria-label="Selected Choice a:"
            className="answer-letter"
            data-test-id="answer-choice-a"
            disabled={true}
            onClick={[MockFunction]}
          >
            a
          </button>
        </span>
        <div
          className="answer-answer"
        >
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "True",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <aside
    className="question-feedback bottom"
  >
    <div
      aria-label="Answer Feedback"
      className="arrow"
    />
    <aside>
      <div
        className="question-feedback-content has-html"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Feedback",
          }
        }
      />
    </aside>
  </aside>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled answer-correct"
      role="region"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-1"
        name="1-options"
        onChange={[MockFunction]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
      >
        <span
          className="answer-letter-wrapper"
        >
          <button
            aria-label="Choice b:"
            className="answer-letter"
            data-test-id="answer-choice-b"
            disabled={true}
            onClick={[MockFunction]}
          >
            b
          </button>
        </span>
        <div
          className="answer-answer"
        >
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "False",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;

exports[`AnswersTable sorts by given ID order 1`] = `
<div
  className="answers-table"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
      role="region"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-0"
        name="1-options"
        onChange={[MockFunction]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
      >
        <span
          className="answer-letter-wrapper"
        >
          <button
            aria-label="Choice a:"
            className="answer-letter"
            data-test-id="answer-choice-a"
            disabled={true}
            onClick={[MockFunction]}
          >
            a
          </button>
        </span>
        <div
          className="answer-answer"
        >
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "False",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
      role="region"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-1"
        name="1-options"
        onChange={[MockFunction]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
      >
        <span
          className="answer-letter-wrapper"
        >
          <button
            aria-label="Choice b:"
            className="answer-letter"
            data-test-id="answer-choice-b"
            disabled={true}
            onClick={[MockFunction]}
          >
            b
          </button>
        </span>
        <div
          className="answer-answer"
        >
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "True",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;
